AMD Geode LX hardware RNG driver broken

### Describe the bug

AMD Geode LX hardware RNG driver is broken since OpenWRT 19.07.

Driver path: drivers/char/hw_random/geode-rng.c

### OpenWrt version

r23389-5deed175a5

### OpenWrt target/subtarget

x86/geode

### Device

Alix board 2d13

### Image kind

Official downloaded image

### Steps to reproduce

1. Install any OpenWRT image later than 19.07.9
2. Open console
3. cat /dev/hwrng

### Actual behaviour

Output in console shows regular pattern (see bellow):

��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��`I��^C

Package rng-tools (rngd daemon) refuses to run because of low entropy from /dev/hwrng

### Expected behaviour

True random output like this:

                     #) R��j���eB�m�c�a@=�ێ�#��E,�0�s�R
ö�2a��n`3t8O(bb�F�	um��^����A�*�p(��`-`�)O���V`�͂JM��	%H����ȴ�i'�Q�s�Nn�����Xj
                                                                               $7�hV�j��yʊBT�I���gh�����Gt������Žjb3C-��T
                                         ڥڝ��C/zw�
                                                  �T(C��Ĥo��-�����*MVח�r�=Z�`_���),��4v^�E���3�1���([��X��@A�׋�+Ex��t��^C


### Additional info

Problem started in OpenWRT releases later than 19.07.9 (last working release).

### Diffconfig

_No response_

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

Having troubles compiling 23.05 branch from git...

 make[1] world
 make[2] tools/compile
 make[3] -C tools/flock compile
 make[3] -C tools/xz compile
 make[3] -C tools/sed compile
 make[3] -C tools/patch compile
 make[3] -C tools/libdeflate compile
 make[3] -C tools/tar compile
 make[3] -C tools/ninja compile
 make[3] -C tools/meson compile
 make[3] -C tools/pkgconf clean-build
 make[3] -C tools/pkgconf compile
    ERROR: tools/pkgconf failed to build.
make -r world: build failed. Please re-run make with -j1 V=s or V=sc for a higher verbosity level to see what's going on
make: *** [/home/timur/Downloads/Temporary/git/openwrt-openwrt-23.05/include/toplevel.mk:232: world] Error 1

--- a/target/linux/x86/patches-5.15/120-hwrng-geode-fix-accessing-registers.patch
+++ b/target/linux/x86/patches-5.15/120-hwrng-geode-fix-accessing-registers.patch
@@ -1,0 +1,47 @@
+From 859bd2e0c0052967536f3f902716f204d5a978b1 Mon Sep 17 00:00:00 2001
+From: Jonas Gorski <jonas.gorski@gmail.com>
+Date: Fri, 8 Sep 2023 22:48:33 +0200
+Subject: [PATCH] hwrng: geode: fix accessing registers
+
+When the membase and pci_dev pointer were moved to a new struct in priv,
+the actual membase users were left untouched, and they started reading
+out arbitrary memory behind the struct instead of registers. This
+unfortunately turned the RNG into a constant number generator, depending
+on the content of what was at that offset.
+
+To fix this, update geode_rng_data_{read,present}() to also get the
+membase via amd_geode_priv, and properly read from the right addresses
+again.
+
+Fixes: 9f6ec8dc574e ("hwrng: geode - Fix PCI device refcount leak")
+Reported-by: Timur I. Davletshin <timur.davletshin@gmail.com>
+Closes: https://bugzilla.kernel.org/show_bug.cgi?id=217882
+Tested-by: Timur I. Davletshin <timur.davletshin@gmail.com>
+Suggested-by: Jo-Philipp Wich <jo@mein.io>
+Signed-off-by: Jonas Gorski <jonas.gorski@gmail.com>
+---
+ drivers/char/hw_random/geode-rng.c | 6 ++++--
+ 1 file changed, 4 insertions(+), 2 deletions(-)
+
+--- a/drivers/char/hw_random/geode-rng.c
++++ b/drivers/char/hw_random/geode-rng.c
+@@ -58,7 +58,8 @@ struct amd_geode_priv {
+ 
+ static int geode_rng_data_read(struct hwrng *rng, u32 *data)
+ {
+-	void __iomem *mem = (void __iomem *)rng->priv;
++	struct amd_geode_priv *priv = (struct amd_geode_priv *)rng->priv;
++	void __iomem *mem = priv->membase;
+ 
+ 	*data = readl(mem + GEODE_RNG_DATA_REG);
+ 
+@@ -67,7 +68,8 @@ static int geode_rng_data_read(struct hw
+ 
+ static int geode_rng_data_present(struct hwrng *rng, int wait)
+ {
+-	void __iomem *mem = (void __iomem *)rng->priv;
++	struct amd_geode_priv *priv = (struct amd_geode_priv *)rng->priv;
++	void __iomem *mem = priv->membase;
+ 	int data, i;
+ 
+ 	for (i = 0; i < 20; i++) {

